### ISSUE TITLE
Prevent user from reloading checkout page while payment is being processed

### ISSUE THREAD

**Issue by asafmjul** [woocommerce/woocommerce#24429]
**Describe the bug**
After a user clicks the “Place Order” button on checkout page, the page is being blocked by an overlay and a loader GIF until the ajax finishes and the user is redirected to Thank You page.

But, if a user reloads the checkout page while the payment is processing, a new order is being created and both of the orders are being charged.

The basic thing is to use the 'beforeunload' event to warn the user before reloading.

The problem is that 'beforeunload' catches also the automatic redirect to Thank You page, and pops alert after the payment was completed successfully and before redirecting to the Thank You page.

I tried two methods -

- Try to recognise specifically a reload with "performance.navigation.type" but this method is not consistent (and not standard anyway).
- Try to unbind the 'beforeunload' after the ajax finishes and before the redirect to the Thank You page - but WooCommerce has no custom event there (somthing like "beforeredirecttothankyou') and I couldn't get to this exact spot with Ajax functions
I'm trying to understand if it can be done, or core code must be overridden for that.

**To Reproduce**
Attached a screen shot

**Screenshots**
https://www.loom.com/share/2e741345d4f848e7ab32d2636b9b5a84

**Expected behavior**
A generic  “Are you sure you want to leave this page” alert when trying to reload or leave the checkout page while payment is being processing.

Another option is just to add a custom JS event handler after the payment Ajax returns, but before the redirect, this will help achieve what we are trying.

**IMPORATNT: The issue is not about the double orders (the creations of the double orders depends on the timing of the reload and not easily reproduced). It is about warning the customer before reloading or leaving the checkout page while a payment is being processed.**

**Isolating the problem (mark completed items with an [x]):**
- [V ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ V] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ V] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
```
`
### WordPress Environment ###

WordPress address (URL): https://jul.co.il
Site address (URL): https://jul.co.il
WC Version: 3.6.5
Log Directory Writable: ✔
WP Version: 5.2.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: –
Language: he_IL
External object cache: –

### Server Environment ###

Server Info: nginx/1.15.12
PHP Version: 7.3.5-1+ubuntu16.04.1+deb.sury.org+1
PHP Post Max Size: 128 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.2.25-MariaDB-10.2.25+maria~xenial-log
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔
Zapier Integration - Active Feeds: ✔ 3 active Zapier Feeds.
Zapier Integration - Cron Tasks: ✔ 0 pending Zapier cron tasks.
Zapier Feed #1 Name: ✔ New Google Contact and Notification Email on New Customer [Universal]
Zapier Feed #1 Webhook URL: ✔ https://hooks.zapier.com/hooks/catch/3672214/qqktkq/
Zapier Feed #1 Trigger: ✔ New Order Status Change
Zapier Feed #1 Test Result: ✔ Successfully sent sample data to https://hooks.zapier.com/hooks/catch/3672214/qqktkq/.
Zapier Feed #2 Name: ✔ New MessageBird Contact on New Customer [Universal]
Zapier Feed #2 Webhook URL: ✔ https://hooks.zapier.com/hooks/catch/3672214/jsbqtc/
Zapier Feed #2 Trigger: ✔ New Order Status Change
Zapier Feed #2 Test Result: ✔ Successfully sent sample data to https://hooks.zapier.com/hooks/catch/3672214/jsbqtc/.
Zapier Feed #3 Name: ✔ Add New Customers to Google Contacts
Zapier Feed #3 Webhook URL: ✔ https://hooks.zapier.com/hooks/catch/3672214/ooz9dk6/
Zapier Feed #3 Trigger: ✔ New Customer
Zapier Feed #3 Test Result: ✔ Successfully sent sample data to https://hooks.zapier.com/hooks/catch/3672214/ooz9dk6/.

### Database ###

WC Database Version: 3.6.5
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 1853.37MB
Database Data Size: 977.48MB
Database Index Size: 875.89MB
wp_woocommerce_sessions: Data: 10.02MB + Index: 0.14MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 35.56MB + Index: 11.52MB
wp_woocommerce_order_itemmeta: Data: 214.78MB + Index: 177.41MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 1.52MB + Index: 0.20MB
wp_woocommerce_payment_tokenmeta: Data: 1.52MB + Index: 3.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 91.63MB + Index: 47.08MB
wp_duplicator_pro_entities: Data: 0.02MB + Index: 0.02MB
wp_duplicator_pro_packages: Data: 0.02MB + Index: 0.02MB
wp_facetwp_index: Data: 5.52MB + Index: 7.02MB
wp_failed_jobs: Data: 0.02MB + Index: 0.00MB
wp_icl_content_status: Data: 0.02MB + Index: 0.02MB
wp_icl_core_status: Data: 0.02MB + Index: 0.02MB
wp_icl_flags: Data: 0.02MB + Index: 0.02MB
wp_icl_languages: Data: 0.02MB + Index: 0.03MB
wp_icl_languages_translations: Data: 0.19MB + Index: 0.09MB
wp_icl_locale_map: Data: 0.02MB + Index: 0.00MB
wp_icl_message_status: Data: 0.02MB + Index: 0.03MB
wp_icl_mo_files_domains: Data: 0.02MB + Index: 0.02MB
wp_icl_node: Data: 0.02MB + Index: 0.00MB
wp_icl_reminders: Data: 0.02MB + Index: 0.00MB
wp_icl_strings: Data: 4.52MB + Index: 4.91MB
wp_icl_string_packages: Data: 0.02MB + Index: 0.00MB
wp_icl_string_pages: Data: 0.02MB + Index: 0.02MB
wp_icl_string_positions: Data: 0.02MB + Index: 0.02MB
wp_icl_string_status: Data: 0.02MB + Index: 0.02MB
wp_icl_string_translations: Data: 2.52MB + Index: 0.52MB
wp_icl_string_urls: Data: 0.02MB + Index: 0.02MB
wp_icl_translate: Data: 0.05MB + Index: 0.02MB
wp_icl_translate_job: Data: 0.02MB + Index: 0.02MB
wp_icl_translations: Data: 5.52MB + Index: 12.06MB
wp_icl_translation_batches: Data: 0.02MB + Index: 0.00MB
wp_icl_translation_status: Data: 2.52MB + Index: 1.52MB
wp_imagify_files: Data: 0.02MB + Index: 0.08MB
wp_imagify_folders: Data: 0.02MB + Index: 0.03MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_mailchimp_carts: Data: 0.02MB + Index: 0.00MB
wp_mc4wp_log: Data: 10.52MB + Index: 0.00MB
wp_options: Data: 5.23MB + Index: 1.02MB
wp_pmxe_exports: Data: 0.08MB + Index: 0.00MB
wp_pmxe_google_cats: Data: 0.39MB + Index: 0.00MB
wp_pmxe_posts: Data: 0.48MB + Index: 0.00MB
wp_pmxe_templates: Data: 0.02MB + Index: 0.00MB
wp_pmxi_files: Data: 0.02MB + Index: 0.00MB
wp_pmxi_history: Data: 0.02MB + Index: 0.00MB
wp_pmxi_images: Data: 2.52MB + Index: 0.00MB
wp_pmxi_imports: Data: 0.16MB + Index: 0.00MB
wp_pmxi_posts: Data: 1.52MB + Index: 0.00MB
wp_pmxi_templates: Data: 0.14MB + Index: 0.00MB
wp_postmeta: Data: 365.81MB + Index: 376.64MB
wp_posts: Data: 41.55MB + Index: 26.63MB
wp_queue: Data: 0.02MB + Index: 0.00MB
wp_swpext_metrics_clicks: Data: 1.52MB + Index: 3.03MB
wp_swpext_metrics_ids: Data: 27.56MB + Index: 44.16MB
wp_swpext_metrics_queries: Data: 2.52MB + Index: 2.52MB
wp_swpext_metrics_searches: Data: 23.56MB + Index: 51.64MB
wp_swp_cf: Data: 0.02MB + Index: 0.05MB
wp_swp_index: Data: 7.17MB + Index: 5.23MB
wp_swp_log: Data: 19.55MB + Index: 29.58MB
wp_swp_tax: Data: 2.20MB + Index: 3.41MB
wp_swp_terms: Data: 0.42MB + Index: 0.59MB
wp_termmeta: Data: 0.11MB + Index: 0.14MB
wp_terms: Data: 0.05MB + Index: 0.03MB
wp_term_relationships: Data: 3.19MB + Index: 2.25MB
wp_term_taxonomy: Data: 0.06MB + Index: 0.03MB
wp_usermeta: Data: 73.61MB + Index: 56.80MB
wp_users: Data: 3.50MB + Index: 4.55MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_product_meta_lookup: Data: 1.33MB + Index: 1.34MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_recommendations: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_session_activity: Data: 1.52MB + Index: 0.00MB
wp_wpml_mails: Data: 4.52MB + Index: 0.00MB

### Post Type Counts ###

acf-field: 139
acf-field-group: 3
attachment: 15122
customize_changeset: 2
jetpack_migration: 2
option: 2
page: 12
post: 12
product: 7188
revision: 796
scheduled-action: 848
shop_coupon: 897
shop_order: 86978
shop_order_refund: 81
store: 119
wc_zapier_feed: 3
wp_block: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (22) ###

ManageWP - Worker: by GoDaddy – 4.9.0
Advanced Custom Fields PRO: by Elliot Condon – 5.8.2
FacetWP - Load More: by FacetWP
LLC – 0.4.3

FacetWP: by FacetWP
LLC – 3.3.11

Imagify: by WP Media – 1.9.6
Loco Translate: by Tim Whitlock – 2.3.0
Mailchimp for WordPress: by ibericode – 4.5.3
Mailgun: by Mailgun – 1.7.1
MailChimp for WordPress - Premium: by ibericode – 4.5.10
Metorik Helper: by Metorik – 1.3.0
SearchWP Live Ajax Search: by Jonathan Christopher – 1.3.1
SearchWP Metrics: by SearchWP
LLC – 1.0.9.2

SearchWP WooCommerce Integration: by SearchWP
LLC – 1.2.1 – Not tested with the active version of WooCommerce

SearchWP: by SearchWP
LLC – 3.0.7

CardCom Payment Gateway: by CardCom LTD – 3.3.1.7 – Not tested with the active version of WooCommerce
WooCommerce Google Analytics Integration: by WooCommerce – 1.4.10
WooCommerce Mixpanel: by SkyVerge – 1.13.1
WooCommerce Smart Coupons: by StoreApps – 4.0.8
WooCommerce Zapier Integration: by OM4 – 1.9.4
WooCommerce: by Automattic – 3.6.5 – 3.7.0 is available
WP All Import Pro: by Soflyy – 4.5.7
WP All Import - WooCommerce Add-On Pro: by Soflyy – 3.0.8

### Inactive Plugins (12) ###

Advanced Custom Fields Multilingual: by OnTheGoSystems – 1.0.1
FacetWP - WPML: by FacetWP
LLC – 1.3.0

SearchWP Custom Results Order: by SearchWP
LLC – 1.1.2

SearchWP Diagnostics: by Jonathan Christopher – 1.4.1
SearchWP WPML Integration: by Jonathan Christopher – 1.4.0
WooCommerce Multilingual: by OnTheGoSystems – 4.6.5
WP All Export - User Add-On Pro: by Soflyy – 1.0.0
WP All Export Pro: by Soflyy – 1.5.6
WPML All Import: by OnTheGoSystems – 2.0.8
WPML Multilingual CMS: by OnTheGoSystems – 4.1.3
WPML String Translation: by OnTheGoSystems – 2.9.1
WPML Translation Management: by OnTheGoSystems – 2.7.2

### Must Use Plugins (2) ###

Kinsta Must-use Plugins: by Kinsta Team – 2.3.1
ManageWP - Worker Loader: by GoDaddy –

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: ILS (₪)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔

### WC Pages ###

Shop base: #5 - /shop/
Cart: #44 - /cart/
Checkout: #45 - /checkout/
My account: #46 - /my-account/
Terms and conditions: #401 - /terms-of-use/

### Theme ###

Name: Storefront Child Theme
Version: 1.0.0
Author URL: http://stuartduff.com
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 2.5.1
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: storefront-child-theme/woocommerce/global/quantity-input.php
storefront-child-theme/woocommerce/product-searchform.php


### Action Scheduler ###

Complete: 802
Oldest: 2019-07-24 15:41:05 +0300
Newest: 2019-08-14 15:56:02 +0300

Pending: 1
Oldest: 2019-08-21 15:56:02 +0300
Newest: 2019-08-21 15:56:02 +0300

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 45
Oldest: –
Newest: –

`
```
</details>

**Comment by claudiosanches** [comment]
Please, feel free to submit a Pull Request to include this behavior.

**Comment by asafmjul** [comment]
Well, unfortunately I'm not experienced enough in git to do it yet. My hands are quite full and I think I wouldn't be able to get around learning it soon.

But I really think these simple lines should be added.

Can I supply the code and some one will submit a Pull Request for it?

**Comment by kloon** [comment]
@asafmjul feel free to attach code here and someone will turn it into a PR. Thanks!

**Comment by JohnTendik** [comment]
Hey, do you guys mind if I take a look at this feature? Due to our slow web servers we're noticing a lot of users are refreshing the page during the submit form stage and ending up with double orders :D I just want to ask and confirm so I'm not stepping on anyone toes. @asafmjul maybe we can collaborate together on this?

**Comment by asafmjul** [comment]
Sure, go ahead. Let me know if you need more info.

**Comment by JohnTendik** [comment]
Modern browsers do not allow for custom messaging during the unload event. We also cannot use confirm, prompt or alert during this event. Would it be better to use defaults and have everyone receive the default dialog message for example in chrome its:

(Reload website?) 
"Changes you made may not be saved"

Its a very generic message. Is it better to go with the generic message, or have some sort of event  after the ajax finishes and before the redirect to the Thank You page?

`$( document.body ).trigger( 'before_thankyou_redirect')`

If we go the event route, that means we're leaving it up to individual shop owners/developer to custom code, bind and unbind the unload event. 

Any thoughts?

**Comment by asafmjul** [comment]
What I tried to do is to bind to the unload event.

The thing is that the bind also worked on a regular successful transaction - so people got the prompt before redirected to the Thank You page on a successful payment.

I couldn't find a way to unbind the event on successful Ajax, so I had to add custom code (and this is why I opened the ticket). But actually, even with the custom code the prompt seems to be affecting more flows that I didn't predicted, so I ended up disabling it for now.

So, as for my research - you have to use the unload event, have to add custom JS events, and have to make modifications to prevent the prompt from interfering other flows.

BTW, what gateway are you using?

**Comment by JohnTendik** [comment]
Hey,

I added a pull request with my proposed changes, please let me know if I should change anything. Also I think we should continue the discussion in the PR (#24609

**Comment by claudiosanches** [comment]
Thank you for reporting this issue!

I'm closing the issue as it already has a PR, please continue the discussion over at #24609.

**Comment by rakosolapov** [comment]
The issue still persists for me.
If I set the network speed limit to 3G in devTools, attempt to place an order, and reload the page (after 5 seconds or less), the page reloads without a message, but the order is still created in WordPress, resulting in duplicate orders (with different IDs and the same cart).
It appears that Woocommerce-stripe-getway calls checkout_place_order_stripe and initially returns false, which skips executing wc_checkout_form.attachUnloadEventsOnSubmit() the first time it runs. A few seconds later, the submit property fires again, and this time the checkout_place_order_stripe trigger returns true, and attachUnloadEventsOnSubmit fires. Perhaps it's worth moving the attachUnloadEventsOnSubmit function higher up so that it executes without regard for the checkout_place_order and checkout_place_order_* handlers?

**Comment by JohnTendik** [comment]
Hey @Maintrust are you able to open a PR for the change? This was very old solution, things may have changed by now. I haven't worked in this codebase in many years so I wont be able to assist with much. I'd recommend opening up a new ticket/bug report and creating a PR to push the fix :)
